### ISSUE TITLE
Optimize the Performance of Bulk Deletion

### ISSUE THREAD

**Issue by scsldb** [pingcap/tidb#18028]
### Description

Optimize the performance and efficiency of bulk deletion, a typical query is:

```sql
delete * from table where id < ?
```

### Category

Feature, Stability

### Value

Implement a special SQL using offset in session to improve performance `delete * from table where t < xxx (backgroud & no-automic)`.

### Time

GanttStart: 2020-07-31
GanttDue: 2020-08-15

**Comment by zhangjinpeng1987** [comment]
The original scenario should be 
```
affect_rows = 1
while affect_rows > 0 {
    affect_rows = `delete * from table where t < xxx limit 5000`
}
```

**Comment by zhangjinpeng1987** [comment]
At first, we can test how about the performance when enabling coprocessor cache.

**Comment by zhangjinpeng1987** [comment]
https://github.com/pingcap/tidb/issues/18308 may be a common solution for this issue.

**Comment by zz-jason** [comment]
@SunRunAway PTAL. If we rely on coprocessor cache to speed up this scenario, we should make coprocessor cache generally available at least in 5.0.0.

BTW, we can also brainstorm whether there are other methods to speed up this scenario.

**Comment by zz-jason** [comment]
> At first, we can test how about the performance when enabling coprocessor cache.

We need to construct a bulk delete scenario to reproduce the performance issue.

**Comment by zz-jason** [comment]
Another solution is to separate all the data into "pages", calculate the page bounds, delete the data "page" by "page":

create a table with `shard_row_id_bits` equals to `4` and load some data:

```sql
create table t(a bigint, b bigint) shard_row_id_bits = 4;
insert into t values(1, 1);
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
```

separate all the data into "pages", calculate the page bounds, in the following example, each "page" contains `5` rows:
```sql
set @batch_size = 5;

select
    min(tmp._tidb_rowid) as start_rowid,
    max(tmp._tidb_rowid) as end_rowid,
    count(*) as page_size
from (
    select
        *,
        _tidb_rowid as _tidb_rowid,
        row_number () over (order by t._tidb_rowid) as row_num
    from t
) tmp
group by floor((tmp.row_num - 1) / @batch_size)
order by start_rowid;
```

delete them page by pages:
```sql
delete from t where _tidb_rowid >=                   5 and _tidb_rowid <=  576460752303423505;
delete from t where _tidb_rowid >=  576460752303423506 and _tidb_rowid <=  576460752303423510;
delete from t where _tidb_rowid >=  576460752303423511 and _tidb_rowid <=  576460752303423515;
delete from t where _tidb_rowid >=  576460752303423516 and _tidb_rowid <=  576460752303423520;
delete from t where _tidb_rowid >= 1152921504606846979 and _tidb_rowid <= 2882303761517117475;
delete from t where _tidb_rowid >= 2882303761517117476 and _tidb_rowid <= 2882303761517117480;
delete from t where _tidb_rowid >= 2882303761517117481 and _tidb_rowid <= 2882303761517117485;
delete from t where _tidb_rowid >= 2882303761517117486 and _tidb_rowid <= 2882303761517117490;
delete from t where _tidb_rowid >= 2882303761517117491 and _tidb_rowid <= 2882303761517117495;
delete from t where _tidb_rowid >= 2882303761517117496 and _tidb_rowid <= 2882303761517117500;
delete from t where _tidb_rowid >= 2882303761517117501 and _tidb_rowid <= 2882303761517117505;
delete from t where _tidb_rowid >= 2882303761517117506 and _tidb_rowid <= 2882303761517117510;
delete from t where _tidb_rowid >= 2882303761517117511 and _tidb_rowid <= 2882303761517117515;
delete from t where _tidb_rowid >= 2882303761517117516 and _tidb_rowid <= 2882303761517117520;
delete from t where _tidb_rowid >= 2882303761517117521 and _tidb_rowid <= 2882303761517117525;
delete from t where _tidb_rowid >= 2882303761517117526 and _tidb_rowid <= 2882303761517117530;
delete from t where _tidb_rowid >= 2882303761517117531 and _tidb_rowid <= 2882303761517117535;
delete from t where _tidb_rowid >= 2882303761517117536 and _tidb_rowid <= 2882303761517117540;
delete from t where _tidb_rowid >= 2882303761517117541 and _tidb_rowid <= 2882303761517117545;
delete from t where _tidb_rowid >= 2882303761517117546 and _tidb_rowid <= 2882303761517117550;
delete from t where _tidb_rowid >= 2882303761517117551 and _tidb_rowid <= 2882303761517117555;
delete from t where _tidb_rowid >= 2882303761517117556 and _tidb_rowid <= 2882303761517117560;
delete from t where _tidb_rowid >= 2882303761517117561 and _tidb_rowid <= 2882303761517117565;
delete from t where _tidb_rowid >= 2882303761517117566 and _tidb_rowid <= 4611686018427387913;
delete from t where _tidb_rowid >= 4611686018427387914 and _tidb_rowid <= 4611686018427387918;
delete from t where _tidb_rowid >= 4611686018427387919 and _tidb_rowid <= 5764607523034235011;
delete from t where _tidb_rowid >= 5764607523034235012 and _tidb_rowid <= 5764607523034235016;
delete from t where _tidb_rowid >= 5764607523034235017 and _tidb_rowid <= 5764607523034235021;
delete from t where _tidb_rowid >= 5764607523034235022 and _tidb_rowid <= 5764607523034235026;
delete from t where _tidb_rowid >= 5764607523034235027 and _tidb_rowid <= 5764607523034235031;
delete from t where _tidb_rowid >= 5764607523034235032 and _tidb_rowid <= 5764607523034235036;
delete from t where _tidb_rowid >= 5764607523034235037 and _tidb_rowid <= 5764607523034235041;
delete from t where _tidb_rowid >= 5764607523034235042 and _tidb_rowid <= 5764607523034235046;
delete from t where _tidb_rowid >= 5764607523034235047 and _tidb_rowid <= 5764607523034235051;
delete from t where _tidb_rowid >= 5764607523034235052 and _tidb_rowid <= 5764607523034235056;
delete from t where _tidb_rowid >= 5764607523034235057 and _tidb_rowid <= 5764607523034235061;
delete from t where _tidb_rowid >= 5764607523034235062 and _tidb_rowid <= 5764607523034235066;
delete from t where _tidb_rowid >= 5764607523034235067 and _tidb_rowid <= 5764607523034235071;
delete from t where _tidb_rowid >= 5764607523034235072 and _tidb_rowid <= 5764607523034235076;
delete from t where _tidb_rowid >= 5764607523034235077 and _tidb_rowid <= 5764607523034235081;
delete from t where _tidb_rowid >= 5764607523034235082 and _tidb_rowid <= 5764607523034235086;
delete from t where _tidb_rowid >= 5764607523034235087 and _tidb_rowid <= 5764607523034235091;
delete from t where _tidb_rowid >= 5764607523034235092 and _tidb_rowid <= 5764607523034235096;
delete from t where _tidb_rowid >= 5764607523034235097 and _tidb_rowid <= 5764607523034235101;
delete from t where _tidb_rowid >= 5764607523034235102 and _tidb_rowid <= 5764607523034235106;
delete from t where _tidb_rowid >= 5764607523034235107 and _tidb_rowid <= 5764607523034235111;
delete from t where _tidb_rowid >= 5764607523034235112 and _tidb_rowid <= 5764607523034235116;
delete from t where _tidb_rowid >= 5764607523034235117 and _tidb_rowid <= 5764607523034235121;
delete from t where _tidb_rowid >= 5764607523034235122 and _tidb_rowid <= 5764607523034235126;
delete from t where _tidb_rowid >= 5764607523034235127 and _tidb_rowid <= 5764607523034235131;
delete from t where _tidb_rowid >= 5764607523034235132 and _tidb_rowid <= 5764607523034235136;
delete from t where _tidb_rowid >= 7493989779944505346 and _tidb_rowid <= 7493989779944505346;
```

with the filters on `_tidb_rowid`, we only need to scan the desired key-value pairs, hope that would save the performance:

```sql
TiDB(root@127.0.0.1:test) > explain delete from t where _tidb_rowid >=                   5 and _tidb_rowid <=  576460752303423505;
+--------------------------+---------+-----------+---------------+--------------------------------------------------------------+
| id                       | estRows | task      | access object | operator info                                                |
+--------------------------+---------+-----------+---------------+--------------------------------------------------------------+
| Delete_4                 | N/A     | root      |               | N/A                                                          |
| └─TableReader_7          | 256.00  | root      |               | data:TableRangeScan_6                                        |
|   └─TableRangeScan_6     | 256.00  | cop[tikv] | table:t       | range:[5,576460752303423505], keep order:false, stats:pseudo |
+--------------------------+---------+-----------+---------------+--------------------------------------------------------------+
3 rows in set (0.00 sec)
```

**Comment by SunRunAway** [comment]
> create a table with `shard_row_id_bits` equals to `4` and load some data:

I think it's a good approaching, what if we provide a command-line tool firstly? After the stable use, we can involve it into SQL command.

**Comment by lzmhhh123** [comment]
I have test continuous delete with coprocessor cache or without coprocessor cache.

The environment is:
- 3 TiDB, 3 PD, 3 TiKV, 3 TiFlash
- version is v4.0.6
- use tpch factor 10, lineitem table

The script is:
```python
#!/usr/bin/env  python
#encoding=utf-8
import mysql.connector

conn = mysql.connector.connect(host='172.16.5.85', port=8092, user='root', password='', database='tpch_10')

for i in range(2000):
    cursor = conn.cursor()
    cursor.execute("delete from lineitem limit 10000")
    print(cursor.rowcount, i)
    conn.commit()
    cursor.close()

print("finished")
```

By monitor of Grafanaz, the result is:

* TiDB with coprocessor cache:
![image](https://user-images.githubusercontent.com/26040259/94098192-aeb79500-fe5a-11ea-99e6-f84e6f7af228.png)

* TiDB without coprocessor cache:
![image](https://user-images.githubusercontent.com/26040259/94098641-aad84280-fe5b-11ea-95b2-65998792fd79.png)


When I turn on the coprocessor cache. P99 is about 250ms, p999 is about 500ms and the jitter is very stable. When turn off the coprocessor cache. P99 is about 1s, p999 is about 2s. By the way, the script got an error `TiKV server timeout` halfway. So in my opinion, turn on the coprocessor cache, to a great extent, could resolve the issue.

**Comment by lzmhhh123** [comment]
For now, we could turn on the copr-cache to solve the problem by temp. If we need some improvement in the future. I will open the issue again.

**Comment by zz-jason** [comment]
@lzmhhh123 could you sumarize the time to delete all the data in the table with or without coprocessor cache?

**Comment by zz-jason** [comment]
What's the coprocessor cache hit ratio when the coprocessor cache is enabled?

**Comment by lzmhhh123** [comment]
> @lzmhhh123 could you sumarize the time to delete all the data in the table with or without coprocessor cache?

I only delete 20M rows data. The number of total rows is 60M. For turning on coprocessor cache, it cost about 6 minutes. For turning off coprocessor cache, it cost 9 min to delete 9M rows and then got an error `TiKV server timeout`.

> * What's the coprocessor cache hit ratio when the coprocessor cache is enabled?

There is not an accurate value now. However, it may exceed 90% I guess.

> * What's the memory usage of coprocessor cache enabled and disabled respectively?

here is the memory usage:
![image](https://user-images.githubusercontent.com/26040259/94108661-45dc1700-fe72-11ea-934a-725424ac3483.png)

* with copr-cache: 11:08-11:14, memory increase from 400MB to 1.1GB
* without copr-cahce: 11:34-11:43, memory is about 250MB+

By the way, the lots of memory cost in middle time is because of load data.

> * Is there any scenario that the coprocessor cache can not speed up the bulk deletion?

Yes, the user manually maintains the deleted range for each iteration.

**Comment by zz-jason** [comment]
> I only delete 20M rows data. The number of total rows is 60M. For turning on coprocessor cache, it cost about 6 minutes.

Could you add more tests to help users know more about the deletion speed on different data sizes and the batch size for each delete transaction?

1. For different records, how about testing 1, 10, and 100 million records?
2. For the best batch size (`N` in `limit N`), how about testing 1000, 10000, 100000, and other batch sizes to find which is best?

Finally, we can add a new [best practice](https://docs.pingcap.com/tidb/stable/tidb-best-practices) document about the bulk deletion.

> There is not an accurate value now. However, it may exceed 90% I guess.

It's better to add some metrics to observe the exact number. BTW, Do we have metrics about the memory consumption of coprocessor cache?

> Yes, the user manually maintains the deleted range for each iteration.

It's important for the users to know more about what scenario can benefit from coprocessor cache and what scenarios can not. Could you summarize these scenarios with some examples and experiments?

**Comment by lzmhhh123** [comment]
@zz-jason Metrics is added, PR is under review. For [best practice](https://docs.pingcap.com/tidb/stable/tidb-best-practices), I will do more experiments and publish an essay for it.
